app/grandchallenge/cases/widgets.py

```diff
@@ -200,3 +205,88 @@ def compress(self, values):
             if len(non_empty_values) != 1:
                 raise ValidationError("Too many values returned.")
             return non_empty_values[0]
+
+
+DICOMUploadWidgetSuffixes = ["dicom-image-name", "dicom-user-uploads"]
+
+
+class DICOMUploadWithName(NamedTuple):
+    name: str
+    user_uploads: list[UserUpload]
+
+
+class DICOMFileNameInput(TextInput):
+    template_name = "cases/file_name.html"
+
+
+class DICOMUploadWidget(MultiWidget):
+    template_name = "cases/dicom_upload_widget.html"
+
+    def __init__(self, attrs=None):
+        # pass in the widgets as dictionary with the proper suffixes as keys,
+        # so we can reliably tell these fields apart from regular socket fields
+        # in forms that include this field
+        widgets = {
+            DICOMUploadWidgetSuffixes[0]: DICOMFileNameInput(),
+            DICOMUploadWidgetSuffixes[1]: DICOMUserUploadMultipleWidget(),
+        }
+        super().__init__(widgets, attrs)
+
+    def decompress(self, value: DICOMUploadWithName):
+        if value:
+            return [
+                value.name,
+                value.user_uploads,
+            ]
+        return [None, None]
+
+
+class DICOMUploadField(MultiValueField):
+    widget = DICOMUploadWidget
+
+    def __init__(self, *args, user, initial, **kwargs):
+        upload_qs = filter_by_permission(
+            queryset=UserUpload.objects.all(),
+            user=user,
+            codename="change_userupload",
+        ).filter(status=UserUpload.StatusChoices.COMPLETED)
+
+        fields = [
+            CharField(),
+            ModelMultipleChoiceField(queryset=upload_qs),
+        ]
+
+        self.current_value = None
+        if initial:
+            # Initial data can only be an image CIV.
+            # We don't want to show the widgets in this case, and instead
+            # display the current image name, so pass the image as
+            # current_value to the widget template
+            if isinstance(initial, ComponentInterfaceValue):
+                self.current_value = initial.image
+                # turn initial to the internal data type that this widget expects
+                initial = DICOMUploadWithName(
+                    name=initial.image.name, user_uploads=[initial.image.pk]
+                )
+            else:
+                raise RuntimeError(
+                    f"Unexpected initial value of type {type(initial)}"
+                )

```

Comment:
This logic could also move to `InterfaceFormfield`. I kept it here so that it aligns with the other custom fields we have, but since those will likely change in the near future, we might want to move all initial handling logic out of the respective fields.